dvc/project.py

```diff
@@ -566,8 +566,15 @@ def _cleanup_unused_links(self, all_stages):
     def checkout(
         self, target=None, with_deps=False, force=False, recursive=False
     ):
-
         if target and not recursive:
+            from dvc.stage import Stage, StageFileDoesNotExistError
```

Comment:
Sorry, didn't communicate good enough my message in https://github.com/iterative/dvc/pull/1580/files#r253734555 . What I meant is that instead of using `if`, we could simply catch both exceptions:

  1) `Stage files should be named 'Dvcfile' or have a '.dvc' suffix(e.g. 'foo.dvc').` (need to create a special exeption for it, since we are currently raising it as DvcException(msg);
  2) `dvc.stage.StageFileDoesNotExistError: 'master' does not exist`

and add a message about "git checkout <branch>". I would consider doing that in `dvc/command/checkout.py`. The reason for dropping `if` in favor of catching exceptions is that we won't need to duplicate any logic detecting and raising those exceptions. What do you think about it? :slightly_smiling_face:

Reply:
So what you mean is (in dvc/command/checkout.py):

```
try:
    self.project.checkout(XXX)
except StageFileDoesNotExistError:
    handle exception
except ImproperStageFileError:
   handle exception
```

Or am I missing your point ?
I've tried the above but it does not work quite work `Error: unexpected error - catching classes that do not inherit from BaseException is not allowed`, probably because or the use of `mock`.

Reply:
So what you mean is (in dvc/command/checkout.py):

```python
try:
    self.project.checkout(XXX)
except DvcException:
    msg = XXX
    logger.error(msg)
    return 1
```

In the setting above, `msg` is a prefix of the error message and is thus not displayed at the correct spot.

We could do:

```python
try:
    self.project.checkout(XXX)
except ImproperStageFileError:
   # this is 1.
   handle exception
except StageFileDoesNotExistError:
    # this is 2.
    handle exception

```
 but it does not work quite work,  probably because of the use of `mock`:
```python
Error: unexpected error - catching classes that do not inherit from BaseException is not allowed
```

Reply:
So what you mean is (in dvc/command/checkout.py):

```python
try:
    self.project.checkout(XXX)
except DvcException:
    msg = 'Warn about git'
    logger.error(msg)
    return 1
```

In the setting above, `msg` is a prefix of the error message and is thus not displayed at the correct spot.

Reply:
I mean that there is no need for that `if` in the `Project.checkout()`, since both of those conditions are checked down the line anyway and we could simply catch those exceptions instead of checking `if os.path.exists(target) or not Stage.is_valid_filename(target)` once again ourselves in Project.checkout(). So, we could do something like this:
```
try:
    self.projcet.checkout(target)
except (StageFileDoesNotExistError, StageFileBadNameError) as exc:
    logger.error(str(exc) + " Maybe you meant 'git checkout {}'?".format(target))
    return 1
```
or maybe, even better, we could do that in Project.checkout():
```
try:
    stages = self._collect(target, with_deps=with_deps)
except (StageFileDoesNotExistError, StageFileBadNameError) as exc:
    raise DvcException(str(exc) + " Maybe you meant git checkout {}".format(target))
```
Sorry for not providing clear examples right away :slightly_smiling_face:

Reply:
That's indeed a much better solution, which in retrospect, looks obvious.
I've also fixed a typo in the `_check_dvc_filename` function

Reply:
Thanks! :slightly_smiling_face: